include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1137,49 +1136,114 @@ struct __early_exit_find_or
+                       __brick_tag, __rngs...);
+
+                // 3. Reduce over group: find __dpl_sycl::__minimum (for the __parallel_find_forward_tag),
+                // find __dpl_sycl::__maximum (for the __parallel_find_backward_tag)
+                // or update state with __dpl_sycl::__any_of_group (for the __parallel_or_tag)
+                // inside all our group items
+                if constexpr (__or_tag_check)
+                    __found_local = __dpl_sycl::__any_of_group(__item_id.get_group(), __found_local);
+                else
+                    __found_local = __dpl_sycl::__reduce_over_group(__item_id.get_group(), __found_local,
+                                                                    typename _BrickTag::_LocalResultsReduceOp{});
+
+                // Set local found state value value to global state to have correct result
+                // - do not check __found_local != __init_value due __result_storage isn't initialized with __init_value on host side.
+                if (__local_idx == 0)
+                {
+                    _AtomicType& __found =
+                        *__result_and_scratch_storage_t::__get_usm_or_buffer_accessor_ptr(__result_acc);
+
+                    // Update global (for all groups) non-atomic state with the found index
+                    __found = __found_local;
+                }
+            });
+    });
 
-    auto __n_groups = oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __wgroup_size);
-    __n_groups = ::std::min(__n_groups, decltype(__n_groups)(__max_cu));
+    // Wait and return result
+    return __future(__event_id, __result_storage).get();
+}
 
-    _PRINT_INFO_IN_DEBUG_MODE(__exec, __wgroup_size, __max_cu);
+// Base pattern for __parallel_or and __parallel_find. The execution depends on tag type _BrickTag.
+template <typename KernelName, bool __or_tag_check, typename _ExecutionPolicy, typename _BrickTag,
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+          typename _Kernel,
+#endif
+          typename _AtomicType, typename _Predicate, typename... _Ranges>
+_AtomicType
+__parallel_find_or_impl_multiple_wgs(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&& __exec,
+                                     _BrickTag __brick_tag,
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+                                     _Kernel& __kernel,
+#endif
+                                     const std::size_t __rng_n, const std::size_t __n_groups,
+                                     const std::size_t __wgroup_size, const _AtomicType __init_value, _Predicate __pred,
+                                     _Ranges&&... __rngs)
+{
+    assert("This device does not support 64-bit atomics" &&
+           (sizeof(_AtomicType) < 8 || __exec.queue().get_device().has(sycl::aspect::atomic64)));
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Starts main work...
```

Comment:
Same comment about removing this since I don't think it adds value to the reader being placed right in the beginning.

Reply:
Removed